### PR TITLE
feature/decide option

### PR DESCRIPTION
**Pull request by steve-si**

### CODE REVIEW

File: Session.js [hunk 1]
```diff
@@ -113,6 +113,33 @@ module.exports = class Session {
...
+   * @param  {Function} cb - error and decision
+   */
+  conditionalDecide(name, contexts = [], options = {}, cb) {
+    options.timeout = options.timeout || this.timeout;
+    options.url = this.amp.domain + this.amp.apiPath + this.amp.key + "/conditionalDecide";
+
+    if (utils.isFunction(arguments[arguments.length - 1])) cb = arguments[arguments.length - 1];
+
+    this.request({
+      event: name,
```
Comment (ajay-si) [inline Session.js]: fill in key, sessionId, userId, decision just like the regular decide request

File: Session.js [hunk 2]
```diff
@@ -113,6 +113,33 @@ module.exports = class Session {
...
+   * @param  {array} contexts - contexts to choose from
+   * @param  {Object} options (optional) - timeout
+   * @param  {Function} cb - error and decision
+   */
+  conditionalDecide(name, contexts = [], options = {}, cb) {
+    options.timeout = options.timeout || this.timeout;
+    options.url = this.amp.domain + this.amp.apiPath + this.amp.key + "/conditionalDecide";
+
+    if (utils.isFunction(arguments[arguments.length - 1])) cb = arguments[arguments.length - 1];
+
```
Comment (ajay-si) [inline Session.js]: check allCandidates.length just like amp.decide above

File: Session.js [hunk 3]
```diff
@@ -113,6 +113,33 @@ module.exports = class Session {
...
+      event: name,
+      contexts: contexts
+    }, options, (err, response, body) => {
+      if (err || (!body || !body.index)) {
+        if(cb) cb(err, body.indexes);
+      } else {
+        if (cb) cb(null, body.indexes, body);
+      }
+    });
+  }
```
Comment (ajay-si) [inline Session.js]: return list of all chosen candidates like in amp.decide?

File: Session.js (hunk 1 above)
Comment (steve-si) [reply Session.js]: So the request that Yuval sent isn't complete?

File: Session.js [hunk 4]
```diff
@@ -113,6 +113,33 @@ module.exports = class Session {
...
+   *
+   * @param  {string} name - name of event
+   * @param  {array} contexts - contexts to choose from
+   * @param  {Object} options (optional) - timeout
+   * @param  {Function} cb - error and decision
+   */
+  conditionalDecide(name, contexts = [], options = {}, cb) {
+    options.timeout = options.timeout || this.timeout;
+    options.url = this.amp.domain + this.amp.apiPath + this.amp.key + "/conditionalDecide";
+
```
Comment (yanpu-si) [inline Session.js]: Check the existence of `contexts` and throw error?

File: Session.js (hunk 1 above)
Comment (ajay-si) [reply Session.js]: What @yuval-si sent embeds the EventRequest - so it has everything that the regular decide has
```type ConditionalEventRequest struct {
    Req      *EventRequest            `json:"request"`
    Event    string                   `json:"event"`
    Contexts []map[string]interface{} `json:"contexts"`
}```

File: Session.js [hunk 5]
```diff
@@ -113,6 +113,33 @@ module.exports = class Session {
     return allCandidates[0];
   }
 
+  /**
+   * conditionalDecide
+   * Decision option to determine decision to take.
+   *
+   * @param  {string} name - name of event
+   * @param  {array} contexts - contexts to choose from
```
Comment (vidya-si) [inline Session.js]: My understanding is that this should be a superset of a regular decide call? Is that not the case. So shouldn't it also include a decide event name and candidates?

**Comment by steve-si** [comment]
Everyone can hold off on reviewing.  I received more requirements and need to make more changes.

**Comment by steve-si** [comment]
Ready for review.

File: Session.js [hunk 6]
```diff
@@ -113,6 +112,62 @@ module.exports = class Session {
...
+    if (contexts.length === 0) throw new Error('Contexts required for conditional decide.');
+
+    options.timeout = options.timeout || this.timeout;
+    options.url = this.amp.domain + this.amp.apiPath + this.amp.key + "/conditionalDecide";
+
+    const { requestSafeCandidates, allCandidates } = this._formatCandidates(candidates);
+
+    if (utils.isFunction(arguments[arguments.length - 1])) cb = arguments[arguments.length - 1];
+    
+    options.limit ? options.limit : allCandidates.length
```
Comment (yanpu-si) [inline Session.js]: Did you miss something here?

File: Session.js [hunk 7]
```diff
@@ -113,6 +112,62 @@ module.exports = class Session {
...
+   *
+   * @param  {string} name - name of event
+   * @param  {array} candidates - variations to choose from
+   * @param  {array} contexts - contexts to choose from
+   * @param  {Object} options (optional) - timeout
+   * @param  {Function} cb - error and decision
+   */
+  conditionalDecide(name, candidates = [], event, contexts = [], options = {}, cb) {
+    if (!event || event === '') throw new Error('Event name required for conditional decide.');
+    if (contexts.length === 0) throw new Error('Contexts required for conditional decide.');
```
Comment (yanpu-si) [inline Session.js]: `if (!Array.isArray(contexts) || contexts.length ===0)`

File: Session.js (hunk 6 above)
Comment (steve-si) [reply Session.js]: Ha.  yep

File: Session.js (hunk 7 above)
Comment (steve-si) [reply Session.js]: 👍

File: Session.js [hunk 8]
```diff
@@ -113,6 +112,62 @@ module.exports = class Session {
...
+      if (err || (!body || !body.indexes)) {
+        // use default
+        if(cb) cb(err, allCandidates[0]);
+      } else {
+        const rankedCandidates = body.indexes.map(index => allCandidates[index]);
+        if (cb) cb(null, rankedCandidates, body);
+      }
+    });
+
+    return allCandidates[0];
```
Comment (ajay-si) [inline Session.js]: shouldn't the return above be a list?

Comment (steve-si) [reply Session.js]: Yes.  Updated for error condition as well.

File: Session.js [hunk 9]
```diff
@@ -113,14 +110,93 @@ module.exports = class Session {
...
+        const result = {};
+        Object.keys(contexts).forEach(key => {
+          result[key] = allCandidates[body.indexes[key][0]];
+        });
+
+        if (cb) cb(null, result, body);
+      }
+    });
+
+    return allCandidates;
```
[the remark this answers is not in the document]
Comment (steve-si) [reply Session.js]: 👍

**Pull request merged**